ca/ca.go

```diff
@@ -218,3 +219,44 @@ func New(log *log.Logger, db *db.MemoryStore) *CAImpl {
 	}
 	return ca
 }
+
+func (ca *CAImpl) CompleteOrder(order *core.Order) {
+	// Check the authorizations - this is done by the VA before calling
+	// CompleteOrder but we do it again for robustness sake.
+	for _, authz := range order.AuthorizationObjects {
+		// Lock the authorization for reading
+		authz.RLock()
+		if authz.Status != acme.StatusValid {
+			return
+		}
+		authz.RUnlock()
+	}
+
+	order.RLock()
+	if order.Status != acme.StatusPending {
+		ca.log.Printf("Error: Asked to complete order %s is not status pending, was status %s",
+			order.ID, order.Status)
+		return
+	}
+	order.RUnlock()
+
+	// Update the order to reflect that we're now processing it
+	order.Lock()
+	defer order.Unlock()
+	ca.log.Printf("Order %s is fully authorized. Ready to issue", order.ID)
+	order.Status = acme.StatusProcessing
+
+	csr := order.ParsedCSR
+	// issue a certificate for the csr
+	cert, err := ca.newCertificate(csr.DNSNames, csr.PublicKey)
+	if err != nil {
+		ca.log.Printf("Error: unable to issue order: %s", err.Error())
+		return
+	}
+	ca.log.Printf("Issued certificate serial %s for order %s\n", cert.ID, order.ID)
+
+	// Update the order to valid status and store a cert ID for the wfe to use to
+	// render the certificate URL for the order
+	order.Status = acme.StatusValid
+	order.CertID = cert.ID
```

Comment:
Yeah; I wrote this comment before the earlier one. Is the certificate stored in memory separately from the order? If so, then yes a pointer to the certificate would be good. Otherwise, DER should work.